docs/cli/ilab-rag-retrieval.md

```diff
@@ -0,0 +1,264 @@
+| chat.rag.context_refinement.enabled | Whether to enable the query refinement request to the LLM | `false` | `--rag-context-refinement` (boolean) |
+| chat.rag.context_refinement.prompt | Prompt template for query refinement | See examples below | `--rag-context-prompt` |
+| chat.rag.retriever.top_k | The maximum number of documents to retrieve | `10` | `--retriever-top-k` |
+| chat.rag.retriever.min_score_threshold | The minimum score threshold for chunks to be considered | `0.5` | `--retriever-min-score-threshold` |
+| chat.rag.retriever.filters | Comma separated key-value pairs filters to restrict the search space | See **Notes** | `--retriever-filters` |
+| chat.rag.prompt | Prompt template for RAG-based queries | See examples below | `--rag-prompt` |
+
+Equivalent YAML document for the newly proposed options:
+```yaml
+chat:
+  rag:
+    enabled: false
+    context_refinement:
+      enabled: true
+      prompt: |
+        You are an assistant helping refine user queries for retrieval from a vector database. 
+        The goal is to extract the most relevant information for answering the user's request. 
+        Rewrite this query to focus on the key concepts and details that will retrieve the most contextually 
+        relevant results from the database. 
+        Make it concise and specific, avoiding conversational context or ambiguity.
+        Return the refined query in a single sentence. Avoid any undesired comment or consideration.
+        User query:
+        {{user_query}}
+        Refined query:
+    retriever:
+      top_k: 10
+      min_score_threshold: 0.5
+      filters: 'url=doc.pdf'
+    prompt: |
+      Given the following information, answer the question.
+      Context:
+      {{context}}
+      Question: {{question}}
+      Answer:
+```
+
+**Notes**:
+* Example of `filters` for a `milvuslite` DB:
+```
+--retriever-filters "url=https://en.wikipedia.org/wiki/Mausoleum_at_Halicarnassus"
+```
+which translates to the following dictionary in the retrieval request (*):
+```json
+{
+   "operator":"==",
+   "field":"url",
+   "value":"https://en.wikipedia.org/wiki/Mausoleum_at_Halicarnassus"
+}
+```
+(*) This is for `milvuslite`, other DB can have a different implementation
+
+### 3.6 References
+* [Haystack-DocumentSplitter](https://github.com/deepset-ai/haystack/blob/f0c3692cf2a86c69de8738d53af925500e8a5126/haystack/components/preprocessors/document_splitter.py#L55)
+* [MilvusEmbeddingRetriever](https://github.com/milvus-io/milvus-haystack/blob/77b27de00c2f0278e28b434f4883853a959f5466/src/milvus_haystack/milvus_embedding_retriever.py#L18)
+
+
+### 3.7 Workflow Visualization
+<!-- https://excalidraw.com/#json=hyDopdXaChbf6TNuLGYE0,WkLQBF84vypx0JwAL5iwzA -->
+Ingestion pipeline:
+![ingestion-mvp](../images/ingestion-mvp.png)
+Chat pipeline enriched by RAG context:
+![rag-chat](../images/rag-chat.png)
+
+### 3.8 Proposed Implementation Stack
```

Comment:
Is there documentation for Feast?  If so, can you post a link here?

Reply:
https://docs.feast.dev/

https://docs.feast.dev/getting-started/quickstart#what-is-feast is probably a good overview

RAG support is alpha and actively being worked on.

Reply:
I see support for indexing and retrieving in the [alpha-vector-database](https://docs.feast.dev/reference/alpha-vector-database) link, but not other aspects of RAG (e.g., response generation).  Is that also being worked, or is the RAG focus specifically indexing and retrieval?

Reply:
It is exclusively focused on indexing and retrieval. Generation (i.e., inference) would be a separate service. In RHOAI it'd be vLLM with KServe.

Reply:
@franciscojavierarceo wrt to the RAG requirement related to ilab, how can we position Feast?
are you envisioning it as a configurable option for the pipeline steps or what else?
- transformation: instead of sdg/docling
- ingestion: instead of configured vectordb and ingestion stack
- retrieval: instead of selected retrieval stack

Reply:
> @franciscojavierarceo wrt to the RAG requirement related to ilab, how can we position Feast? are you envisioning it as a configurable option for the pipeline steps or what else?

I'd position Feast as basically handling the local experience to ingest and transform data nicely for us and to enable us to scale InstructLab RAG to RHOAI.

And to your specific items:
> 
> * transformation: instead of sdg/docling
Docling is a particular transformation applied to some input data that returns a particular set of output data. At the moment, this is done on some set of batch data. In live production use cases these documents likely exist in a production service that may change and require re-processing upon changes (i.e., ingesting individual documents and processing and persisting them either synchronously or asynchronously). SDG is retrieval + inference in the sense that the retrieved documents are injected in the context before running inference. Even if the retrieval is on batch data, there is important metadata that should be stored during the processing (e.g., if embeddings are used, which model, version of code, etc.). 
> * ingestion: instead of configured vectordb and ingestion stack
Feast has the right primitives for ingestion and the flexibility to support Batch, Streaming, and Request driven data. 

> * retrieval: instead of selected retrieval stack
Again, feast's `get_online_features` or `retrieval_online_documents` apis are good defaults already and the benefits here are that ilab can set the best defaults for ilab using Feast without having to implement them.

I can commit to upstreaming things we feel are a benefit to Feast, the point I'm trying to highlight with all of this is that we will find ourselves having to reinvent many if not all of the same details that a feature store does because ultimately prompts, documents, and chat history (more generally tokens) are just features into a model. Since RHOAI is investing into Feast, it is a natural consideration that doesn't lock us into a single VectorDB and doesn't expose us to conflicts with companies changing the license as it is part of the LF AI and Data project.

Let me know if that makes sense.

Reply:
**TL;DR**: Probably doable. Is it the top priority?

- Yes, if we design the solution with the required flexibility, Feast could be plugged as one of the embedding stores  or even retriever, as you said. Something like:
```
ilab data ingest --store-type feast --store-uri 'http://localhost:875/' \
  --store-settings 'feature_name=AAA,feature=BBB' /path/to/processed/files
ilab model chat --store-type feast --store-uri='...' ...
```
- Related functions should be defined in a dedicated package like `instructlab-feast`, considering the (non-trivial) manipulations required to adapt the two environments.
- IMO, it seems a bit excessive to install and configure Feast in the CLI environment where ilab lives. Maybe this is more relevant for a production environment in RHOAI.
- The outcome for time being is that we could try to be a bit more generic in the adopted terminology. E.g. user `store` instead of `vectordb` when we reference parameter names. This can also come in handy when we integrate with embedding APIs rather than using our own embedder. @jwm4 WDYT?

Reply:
When we get to the RHOAI solution we would benefit from all of the operator, RBAC, integration work we've already done, which is where my view comes from.

If we were to go down that route, ilab data ingest would just be applications of Feast with our ILab specific business logic. I think that'd be a reasonable pattern. 

I do acknowledge though that it could be bloated for what we're doing and Feast RAG is not as strong of a community as the other nor as good of a solution. The big benefit is that Feast is not single-entity backed.

These are all things I think are worth discussing in a doc (or maybe this ADR).

Reply:
For the sake of understanding, could you clarify which services will become `ilab data ingest` and `ilab model chat` when we move to RHOAI?

Reply:
`ilab data ingest` would in an ideal world would be Feast.

`ilab model chat` I'm not sure. It would require be KServe, _potentially_ Feast (or whatever document serving solution we choose), and whatever chat application/service solution we choose.

Reply:
@franciscojavierarceo writes:
> When we get to the RHOAI solution we would benefit from all of the operator, RBAC, integration work we've already done, which is where my view comes from.
> If we were to go down that route, ilab data ingest would just be applications of Feast with our ILab specific business logic. I think that'd be a reasonable pattern.
> I do acknowledge though that it could be bloated for what we're doing and Feast RAG is not as strong of a community as the other nor as good of a solution. The big benefit is that Feast is not single-entity backed.
> These are all things I think are worth discussing in a doc (or maybe this ADR).

These seem like very important topics to me.  I would rather have a separate doc for them rather than try to get it all sorted out in this doc because they also seem very complicated.  FWIW, I think I only understand bits and pieces of the vision here and not the big picture yet, but it does seem like there is a lot of value that this integration could provide for enterprise scale-out.

Reply:
> Can you find other examples too?
For now we don't have many configurations exposed, and `document-store` can also apply to the chat pipeline.

Post-mvp, we could include the `retriever` settings to specify the retriever details:
- `retriever-type`: 
  - `default`: our default retriever implementation using the configured document-store.
  - <another available retriever OOTB>.
  - `api: `a dedicated service exposing well-defined APIs for retrieving embeddings (/query). Includes the document store.
  - `feast`: includes the document store.
Before going into the hell of overthinking, I would wait for the outcome of ET experiments and also carefully evaluate the solution designed by the IBM research team.

Reply:
> `ilab data ingest` would in an ideal world would be Feast.
> 
> `ilab model chat` I'm not sure. It would require be KServe, _potentially_ Feast (or whatever document serving solution we choose), and whatever chat application/service solution we choose.

Agree. That being said, how can this ADR support the request to integrate with Feast?